ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -137,23 +158,36 @@ toBase32 dat = map word5 $ runIdentity $ convertBits (map fromIntegral dat) 8 5
 toBase256 :: [Word5] -> Maybe [Word8]
 toBase256 dat = fmap (map fromIntegral) $ convertBits (map fromWord5 dat) 5 8 noPadding
 
-segwitCheck :: Word8 -> Data -> Bool
-segwitCheck witver witprog =
-    witver <= 16 &&
+segwitCheck :: Word8 -> Data -> Maybe Word
+segwitCheck witver witprog = do
+    guard $ witver <= 16
     if witver == 0
-    then length witprog == 20 || length witprog == 32
-    else length witprog >= 2 && length witprog <= 40
+    then guard (length witprog == 20 || length witprog == 32) >> return 1
+    else guard (length witprog >= 2 && length witprog <= 40) >> return bech32mConst
+ where
+  bech32mConst = 0x2bc830a3
```

Comment:
Might be nice to define a global sum type for the two encoding types and have a mapping function from encoding to checksum constant. Then it can be exported to the tests.

Reply:
Done.